### PR TITLE
*: Avoid store plaintext password in mysql.user table.

### PR DESCRIPTION
**Pull request by shenli**
Store hashed password.

### CODE REVIEW

File: util/auth.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+func EncodePassword(pwd string) string {
+	hash := Sha1Hash([]byte(pwd))
+	return hex.EncodeToString(hash)
+}
+
+// DecodePassword converts hex string password to byte array.
+func DecodePassword(pwd string) ([]byte, error) {
+	x, err := hex.DecodeString(pwd)
+	if err != nil {
+		return []byte{}, errors.Trace(err)
```
Comment (ngaut) [inline util/auth.go]: Why not just return nil  instead of []byte{}

Comment (shenli) [inline util/auth.go]: PTAL

**Comment by kevincox** [comment]
Shouldn't some salt be used? Or ideally a password hashing library so that someone with cryptographic knowledge can maintain the hashing.

File: util/auth.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import (
+	"crypto/sha1"
+	"encoding/hex"
+
+	"github.com/juju/errors"
+)
+
+// CalcPassword is the algorithm convert hashed password to auth string.
+// See: https://dev.mysql.com/doc/internals/en/secure-password-authentication.html
+// SHA1( password ) XOR SHA1( "20-bytes random data from server" <concat> SHA1( SHA1( password ) ) )
```
Comment (qiuyesuifeng) [inline util/auth.go]: Seems `SHA1( SHA1( password ) )` should be `SHA1( password )`  in comment?

File: util/auth.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+// EncodePassword converts plaintext password to hashed hex string.
+func EncodePassword(pwd string) string {
+	hash := Sha1Hash([]byte(pwd))
+	return hex.EncodeToString(hash)
+}
+
+// DecodePassword converts hex string password to byte array.
+func DecodePassword(pwd string) ([]byte, error) {
```
Comment (qiuyesuifeng) [inline util/auth.go]: Can DecodePassword only get hash string not origin password?

**Comment by shenli** [comment]
@kevincox The salt is a random data generated for each client connection.  The parameter scramble of CalcPassword function in auth.go is the salt.
MySQL client uses server-generated salt and password to caculate a cryptographic authstring. Server use the same algorithm and compare check if the result is the same with client.

File: util/auth.go (hunk 3 above)
Comment (shenli) [inline util/auth.go]: Hashed string. There is no way to get the origin password.

File: util/auth.go (hunk 2 above)
Comment (shenli) [inline util/auth.go]: In the comment "password" is the origin password. Maybe I should change the name of the parameter.

Comment (shenli) [inline util/auth.go]: PTAL @qiuyesuifeng

**Comment by siddontang** [comment]
LGTM

**Comment by kevincox** [comment]
Ok, I see where the salt comes in but I'm still unconvinced about the overall system. It appears that there is no way to upgrade the hashes in the future and SHA1 is already getting chipped away at. Also three hashes on SHA1 is very very cheep so it won't do much to stop brute force attacks.

**Pull request merged**